torch/_guards.py

```diff
@@ -290,6 +336,41 @@ def restore_graphstate(self, state):
         self.dynamo_guards = state.dynamo_guards
 
 
+class ModuleContext(Checkpointable[ModuleCheckpointState]):
+    def __init__(self):
+        self.names_to_sources: Dict[str, Source] = dict()
+
+    def copy_graphstate(self):
+        return ModuleCheckpointState(dict(self.names_to_sources))
+
+    def restore_graphstate(self, state):
+        assert isinstance(state, ModuleCheckpointState)
+        self.names_to_sources = state.names_to_sources
+
+    def register(self, source: Source):
+        """
+        Registers a source object to it's given name.
+
+        :param source: A Source object representing the source to be registered.
+        :type source: Source
+
+        :raises AssertionError: If a source with the same name already exists but has a different source name.
```

Comment:
This still seems misdesigned. If I'm reading this correctly, this structure tells you what the source of any named buffers/parameters on the GraphModule being built are. There should never be conflicts, because for every distinct source we should have explicitly allocated a fresh name for them.

Reply:
There are never conflicts, and we never try catch this anywhere, but I wanted a way of detecting potentially illegal state.

Let me make this far more clear.

Reply:
Normalize_attr_name is not a mangler. You can't guarantee distinct sources won't conflict

Reply:
Understood. You said this before, but it did not click till now.

What if we had normalize_attr_name take the tracing context's name_to_sources, and collisions would always increment? I think that might work, right? Let me try a version of this PR that does that.

Reply:
This still sounds structurally wrong. *Somewhere* you're doing name allocation, right? Wherever that is, that should be the one source of truth of names, and then you plumb the name along to wherever you need it otherwise.

Reply:
Name for eval and name for lookup cannot be the same. So, no, we cannot just have one name.